add googleStorage lib wrapper around new google storage client

Qi77Qi opened the pull request:
wip to gather some thoughts.
Motivation
1. Current google library is no longer in active development. https://github.com/googleapis/google-api-java-client#library-maintenance
2. I'd like a google lib exposes `IO` interface instead of only `Future`
3. New storage library has a [java emulator](https://github.com/googleapis/google-cloud-java/blob/master/TESTING.md#testing-code-that-uses-storage). It doesn't cover all features though. 
4. rename the `Ops` class to `Alg` just so it's more consistent with most tagless final implementations

Only implemented methods needed in https://github.com/broadinstitute/sam/blob/develop/src/main/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleKeyCache.scala

TODO:
* check out logging for grpc calls
* https://github.com/broadinstitute/workbench-libs/pull/188#discussion_r237731810

**PR checklist**
- [ ] For each library you've modified here, decide whether it requires a major, minor, or no version bump. (Click [here](/broadinstitute/workbench-libs/blob/develop/CONTRIBUTING.md) for guidance)

If you're doing a **major** or **minor** version bump to any libraries:
- [ ] Bump the version in `project/Settings.scala` `createVersion()`
- [ ] Update `CHANGELOG.md` for those libraries
- [ ] I promise I used `@deprecated` instead of deleting code where possible

In all cases:
- [ ] Replace the appropriate version hashes in `README.md` and the `CHANGELOG.md` for any libs you changed with `TRAVIS-REPLACE-ME` to auto-update the version string
- [ ] Get two thumbsworth of PR review
- [ ] Verify all tests go green (CI _and_ coverage tests)
- [ ] Squash commits and **merge to develop**
- [ ] Delete branch after merge

Qi77Qi committed:
add googleStorage lib wrapper around new google storage client

dvoet commented on google/src/main/scala/org/broadinstitute/dsde/workbench/google/GoogleFirestoreAlg.scala:
```diff
@@ -5,7 +5,7 @@ import com.google.cloud.firestore.{DocumentSnapshot, Firestore, Transaction}
 import scala.language.higherKinds
 
-trait GoogleFirestoreOps[F[_]] {
+trait GoogleFirestoreAlg[F[_]] {
```
What is Alg short for? Can you avoid the abbreviation?

I am guessing it is Algebra but spelling that out is probably more confusing.. How about `GoogleFirestoreService`?

dvoet replied:
clarification: Algebra is not a confusing term, it is quite precise. And while this may be an algebra, I don't think it conveys what we want. 

This is the thing that knows how to talk to Firestore. In the past we have used the term data access object (DAO). I don't really care if we want to use a different term but it should be something that indicates what it does instead of what it is.

Qi77Qi replied:
wouldn't it be more confusing if we use some custom name?
With `Alg`, we can refer people to https://typelevel.org/cats-tagless/, and it's easier for them to follow the pattern.
Maybe I should link https://typelevel.org/cats-tagless/ if that helps? most blog posts will refer this to `Alg` or `Algebra` I think. I'm fine with `Algebra` too if that's better

dvoet commented on google/src/main/scala/org/broadinstitute/dsde/workbench/google/GoogleStorageAlg.scala:
```diff
@@ -0,0 +1,28 @@
+trait GoogleStorageAlg[F[_]] {
+  def listObjectsWithPrefix(bucketName: GcsBucketName, objectNamePrefix: String): Stream[F, GcsObjectName]
+
+  // not memory safe. Use listObjectsWithPrefix if you're worried about OOM
```
put this in scaladoc

Qi77Qi committed:
fix typo

Qi77Qi committed:
drop 2.11 for all modules except util and model since orch depends on them

dvoet commented on google/src/test/scala/org/broadinstitute/dsde/workbench/google/GoogleStorageInterpreterSpec.scala:
```diff
@@ -0,0 +1,99 @@
+    val objectName = genGcsBlobName.sample.get
+    val objectBody = genGcsObjectBody.sample.get
+    val res = localStorage.storeObject(bucketName, objectName, objectBody, objectType).attempt.map(x => x.isRight shouldBe(true))
+    res.unsafeToFuture()
```
I have been getting grumpy about having to call `res.unsafeToFuture` at the end of every test. It is so easy to forget and not even notice. Is there some way we can get scalatest to know about IO? Or maybe we just have a utility function:
```
def ioTest[T](test: => IO[T]) = test.unsafeToFuture()
```
then tests can look like
```
"ioStorage storeObject" should "be able to upload an object" in ioTest {
```

Qi77Qi replied:
I don't think scalatest will know about IO since it doesn't depends on cats-effect.
I think it'll error if you have `IO[Assertion]` at the end of each test instead of `Future[Assertion]`.
I'll create the ioTest helper if you prefer

dvoet commented on google/src/main/scala/org/broadinstitute/dsde/workbench/google/GoogleStorageInterpreters.scala:
```diff
@@ -0,0 +1,89 @@
+    implicit cs: ContextShift[IO]
+  ): GoogleStorageAlg[IO] = new GoogleStorageAlg[IO] {
+    override def listObjectsWithPrefix(bucketName: GcsBucketName, objectNamePrefix: String): Stream[IO, GcsObjectName] = for {
+        pages <- Stream.eval(cs.evalOn(blockingEc)(IO(db.list(bucketName.value, BlobListOption.prefix(objectNamePrefix), BlobListOption.pageSize(maxPageSize), BlobListOption.currentDirectory()))))
```
I don't think this handles queries with results that span pages. This only gets the first page and never subsequent pages.

Qi77Qi replied:
from doc
```
  /**
   * Returns an iterable that traverses all of the elements of the underlying data source. The data
   * is fetched lazily page by page, where each page may contain multiple elements. A new page is
   * fetched whenever the elements of any particular page are exhausted.
   */
  Iterable<ResourceT> iterateAll();
```

dvoet commented on google/src/test/scala/org/broadinstitute/dsde/workbench/google/GoogleStorageInterpreterSpec.scala:
```diff
@@ -0,0 +1,99 @@
+import scala.concurrent.ExecutionContext
+
+// AsyncFlatSpec currently doesn't work with scalacheck's forAll. It'll be supported in scalatest 3
+class GoogleStorageInterpreterSpec extends AsyncFlatSpec with Matchers {
```
add a test where `listObjectsWithPrefix` has more results than fit on a single page, suggest make `maxPageSize` small

Qi77Qi committed:
add retry and address comments

Qi77Qi committed:
make sure listPrefix is listing multiple pages

marctalbott reviewed (approved):
Looks fine, although I can't say I fully understand the fs2 streams and all the functions that go along with them like `compile`, `drain`, `foldMonoid`, etc.

Qi77Qi committed:
rename algebra to GooglexxService

Qi77Qi committed:
don't cross build

Qi77Qi committed:
forget about dropping 2.11 for now

Qi77Qi committed:
respect crossBuildVersions

Qi77Qi committed:
fix travis

Qi77Qi committed:
fix

Qi77Qi committed:
use one version in travis

use toList

use F in interpreters so that we don't have to create a new for for each effect type

use Stream.retry in DistributedLock

address comments

remove unfold and add jitter

fix jitter

bump fs2 version

don't use iterateAll and add back multi page test

add blockingF helper

update sbt version and make unit tests test data more unique

bump scala and google cloud storage version

address some comments

upgrade google storage version

move new google libraries to separate module

remove some old code

fix path

Qi77Qi commented:
Since there has been many times code compiles fine in 2.12 but not in 2.11. I feel we should just drop 2.11 for modules that no projects depends on.

There's some change related to travis
1. replace `++$TRAVIS_SCALA_VERSION` with `+` so that all sbt commands will cross build based on `crossScalaVersions` in sbt
2.  remove 2.11 in `.travis.yml` since we're already building both versions in one build

Qi77Qi commented on google/src/main/scala/org/broadinstitute/dsde/workbench/google/GoogleStorageInterpreters.scala:
```diff
@@ -0,0 +1,124 @@
+  override def listObjectsWithPrefix(bucketName: GcsBucketName, objectNamePrefix: String, maxPageSize: Long = 1000): Stream[IO, GcsObjectName] = {
+    for {
+      firstPage <- retryStorageIO(cs.evalOn(blockingEc)(IO(db.list(bucketName.value, BlobListOption.prefix(objectNamePrefix), BlobListOption.pageSize(maxPageSize.longValue()), BlobListOption.currentDirectory()))))
+      page <- Stream.unfold(firstPage){
```
`Stream.unfold` will keep running the effects until `None`

Qi77Qi commented on .travis.yml:
```diff
@@ -2,7 +2,6 @@ jdk:
 language: scala
 scala:
-- 2.11.12
 - 2.12.7
```
removing `2.11.12` here, but all sbt commands are running with `+`, so it'll cross build respecting `crossScalaVersions` in sbt definition.

Qi77Qi commented on google/src/main/scala/org/broadinstitute/dsde/workbench/google/GoogleStorageInterpreters.scala:
```diff
@@ -0,0 +1,124 @@
+      .toBuilder
+      .setLifecycleRules(lifecycleRules.asJava)
+      .build()
+    retryStorageIO(cs.evalOn(blockingEc)(IO(db.update(bucketInfo)))).compile.drain
```
`.compile.drain` is for running the stream just for effects.
For instance:
```
qi-sam@ Stream.eval(IO(println(1))) >>  Stream.eval(IO(println(2)))
res3: Stream[IO[x], Unit] = Stream(..)

qi-sam@ res3.compile.drain.unsafeRunSync()
1
2

```

`>>` is the same as `flatMap(_ => secondStream)`. The operator exists for all monads once you import `cats.implicits._`

Qi77Qi commented on google/src/main/scala/org/broadinstitute/dsde/workbench/google/GoogleStorageInterpreters.scala:
```diff
@@ -0,0 +1,124 @@
+      blob <- OptionT(retryStorageIO(getBlobs))
+      r <- OptionT.liftF(Stream.emits(blob.getContent()).covary[IO].through(text.utf8Decode))
+    } yield r
+    res.value.compile.foldMonoid
```
`.compile.foldMonoid` over `Option` monoid instance.
`fold` is for reducing stream down to a value.
`foldMonoid` means reducing the stream down to a pure value with monoid's `+` function

Qi77Qi commented on google/src/main/scala/org/broadinstitute/dsde/workbench/google/GoogleStorageInterpreters.scala:
```diff
@@ -0,0 +1,124 @@
+  override def removeObject(bucketName: GcsBucketName, blobName: GcsBlobName): IO[Boolean] = {
+    val deleteObject = cs.evalOn(blockingEc)(IO(db.delete(BlobId.of(bucketName.value, blobName.value))))
+    for {
+      deleted <- retryStorageIO(deleteObject).compile.lastOrError
```
`.compile.lastOrError` will raise error if the stream is empty

Qi77Qi commented on google/src/main/scala/org/broadinstitute/dsde/workbench/google/GoogleFirestoreInterpreter.scala:
```diff
@@ -0,0 +1,85 @@
+import scala.concurrent.ExecutionContext
+
+private[google] class GoogleFirestoreInterpreter[F[_]](db: Firestore)(
+              implicit effect: Effect[F],
```
not sure why git doesn't recognize this as a renamed file..but there's no significant change except I changed constraint to `Effect`

Qi77Qi replied:
use `Effect` constraint, because I need to `unsafeRunSync` on the code that's submitted to google's callback api

rtitle commented on google/src/main/scala/org/broadinstitute/dsde/workbench/google/GoogleStorageInterpreter.scala:
```diff
@@ -0,0 +1,120 @@
+  timer: Timer[F],
+  asyncF: Async[F]
+) extends GoogleStorageService[F] {
+  private def retryStorageF[A](ioa: F[A]): Stream[F, A] = Stream.retry(ioa, retryConfig.retryInitialDelay, retryConfig.retryNextDelay, retryConfig.maxAttempts, retryConfig.retryable)
```
This retry stuff isn't really `Storage`-specific. It'd be cool to move this to a common location in case other Google interpreters want to use the streaming pattern as an alternative to https://github.com/broadinstitute/workbench-libs/blob/develop/util/src/main/scala/org/broadinstitute/dsde/workbench/util/Retry.scala.

Qi77Qi replied:
I don't think it'll actually save much code by extracting this out. But I added 
```
  def retryGoogleF[F[_]: Timer: RaiseThrowable, A](retryConfig: RetryConfig)(fa: F[A]): Stream[F, A] = Stream.retry[F, A](fa, retryConfig.retryInitialDelay, retryConfig.retryNextDelay, retryConfig.maxAttempts, retryConfig.retryable)

```
in package object.

Not sure if that's better or worth though

rtitle commented on google/src/main/scala/org/broadinstitute/dsde/workbench/google/GoogleStorageInterpreter.scala:
```diff
@@ -0,0 +1,120 @@
+        currentPage =>
+          Option(currentPage).map(p => (p, p.getNextPage))
+      }
+      objects <- Stream.fromIterator[F, Blob](page.iterateAll().iterator().asScala).map{
```
It looks like `iterateAll` returns all the elements, _on all the pages_ (e.g. it lazily loads a new page when the current page is exhausted). Since we're already `unfold`ing pages, I wonder if `getValues()` is the method we want to call here, which only operates on the current page.

See: https://github.com/googleapis/gax-java/blob/master/gax/src/main/java/com/google/api/gax/paging/Page.java

Qi77Qi replied:
I thought `iterateAll` will retrieve all pages in the beginning too. But I wrote a test https://github.com/broadinstitute/workbench-libs/pull/188/files#diff-405a305f0d21ea56fc302bdd7467022bR87, and with storage emulator, it only returned one page. That's why I added `unfold` to retrieve all pages

rtitle replied:
Interesting, ok. Maybe it's worth a test with a real bucket with multiple pages? I'm just worried that if `iterateAll` works as advertised you might end up with duplicate entries by calling `iterateAll` for each page.

For example if you have:
```
page1
  obj1
  obj2
page2
  obj3
  obj4
```
would we get
```
Stream(
  /* iterateAll on page1 */ obj1, obj2, obj3, obj4, 
  /* iterateAll on page2 */ obj1, obj2, obj3, obj4)
```
?

I could be completely misunderstanding though..

Qi77Qi replied:
oh I see...I'll try this out tomorrow.

rtitle commented on google/src/main/scala/org/broadinstitute/dsde/workbench/google/GoogleStorageInterpreter.scala:
```diff
@@ -0,0 +1,120 @@
+      firstPage <- retryStorageF(cs.evalOn(blockingEc)(asyncF.delay(db.list(bucketName.value, BlobListOption.prefix(objectNamePrefix), BlobListOption.pageSize(maxPageSize.longValue()), BlobListOption.currentDirectory()))))
+      page <- Stream.unfold(firstPage){
+        currentPage =>
+          Option(currentPage).map(p => (p, p.getNextPage))
```
Does `p.getNextPage()` make a Google API call? If so I'm wondering if it should be retried like the initial call to `list()`.

rtitle commented on google/src/main/scala/org/broadinstitute/dsde/workbench/google/GoogleStorageInterpreter.scala:
```diff
@@ -0,0 +1,120 @@
+    } yield objects
+  }
+
+  override def unsafeGetObject(bucketName: GcsBucketName, blobName: GcsBlobName): F[Option[String]] = {
```
Should we also define a safe `getObject` which returns the object contents as a `Stream`? Would this return a `Stream[F, Byte]`, or could it be chunked somehow (I'm not remembering what fs2 supports here)?

Qi77Qi replied:
if we want to be memory safe, I think we'll need to write to a file. Google storage doesn't expose an api that returns a memory safe data type.

I think we should definitely add this at some point, but since this PR is more for supporting https://github.com/broadinstitute/sam/pull/249 at the moment, which is just retrieving a key file. I think we can add it in a different PR.

rtitle replied:
Gotcha, makes sense.

rtitle commented on google/src/main/scala/org/broadinstitute/dsde/workbench/google/GoogleStorageInterpreter.scala:
```diff
@@ -0,0 +1,120 @@
+  }
+
+  override def removeObject(bucketName: GcsBucketName, blobName: GcsBlobName): F[Boolean] = {
+    val deleteObject = cs.evalOn(blockingEc)(asyncF.delay(db.delete(BlobId.of(bucketName.value, blobName.value))))
```
This pattern is everywhere, I wonder if a helper function would be useful?
```
  def liftAsync[A](thunk: => A): F[A] = {
    cs.evalOn(blockingEc)(asyncF.delay(thunk))
  }
```
Or is that more confusing? I dunno..

Qi77Qi replied:
I kinda prefer to have specific `evalOn` calls since we should be cautious about shifting to blockingEc

rtitle commented on google/src/main/scala/org/broadinstitute/dsde/workbench/google/GoogleStorageInterpreter.scala:
```diff
@@ -0,0 +1,120 @@
+    x => x * 2,
+    5,
+    {
+      case e: StorageException => if(storageRetryableCode.contains(e.getCode)) true else false
```
Nit:
```
case e: StorageException => storageRetryableCode.contains(e.getCode)
```

Qi77Qi replied:
I have no idea why I did `if someBoolean true else false` lol

rtitle commented on google/src/main/scala/org/broadinstitute/dsde/workbench/google/GoogleStorageService.scala:
```diff
@@ -0,0 +1,41 @@
+
+  def storeObject(bucketName: GcsBucketName, objectName: GcsBlobName, objectContents: Array[Byte], objectType: String): F[Unit]
+
+  def setBucketLifecycle(bucketName: GcsBucketName, lifecycleRules: List[LifecycleRule]): F[Unit]
```
Do we care about using `com.google.cloud.storage._` classes in this interface? Should we replicate everything in `workbench-model`? I'm honestly not sure. Right now it looks like we use the Google classes for `Acl` and `LifecycleRule`, whereas everything else (like `GcsBucketName`, `GcsObjectName`) comes from workbench-model.

Qi77Qi replied:
I think exposing google's data types directly gives more options to callers. They're just `String`s, so I think they're fine. Our current model case classes are actually less ideal because google's library api changed. Another reason I don't quite like current google models is there's some json code tied with them, which shouldn't be necessary anymore since new library uses grpc

rtitle commented on google/src/main/scala/org/broadinstitute/dsde/workbench/google/GoogleStorageService.scala:
```diff
@@ -0,0 +1,41 @@
+  def createBucket(billingProject: GoogleProject, bucketName: GcsBucketName, acl: List[Acl]): F[Unit]
+}
+
+final case class GcsBlobName(value: String) extends AnyVal
```
This should maybe go in workbench-model along with the other Gcs model classes.

Qi77Qi replied:
I see wb-model as firecloud/wb related model. Say a new service might want to bring in wb-models, but it doesn't care about calling google, and they probably don't care about google related models.

rtitle commented on google/src/main/scala/org/broadinstitute/dsde/workbench/util/DistributedLockAlgebra.scala:
```diff
@@ -0,0 +1,8 @@
+import org.broadinstitute.dsde.workbench.google.util.LockPath
+
+trait DistributedLockAlgebra[F[_]] {
+  def withLock(lock: LockPath): Resource[F, Unit]
```
Cool, nice use of cats-effect `Resource` to implement a lock. So the caller can do something like this:
```
distributedLock.withLock(LockPath(...)).use(_ => doSomething())
```
and it will automatically acquire and release the lock, correct?

Qi77Qi replied:
yep

rtitle commented on google/src/main/scala/org/broadinstitute/dsde/workbench/util/DistributedLock.scala:
```diff
@@ -38,14 +39,14 @@ class DistributedLock[F[_]](lockPathPrefix: String, config: DistributedLockConfi
         _ <- lockStatus match {
           case Available => setLock(tx, docRef, lock.expiresIn)
-          case Locked => af.raiseError[Unit](new WorkbenchException(s"can't get lock: $lock"))
+          case Locked => asyncF.raiseError[Unit](new WorkbenchException(s"can't get lock: $lock"))
```
So it fails fast if the lock can't be acquired, as opposed to blocking (up to some timeout) waiting for the lock to be available?

Qi77Qi replied:
this function is just for acquiring a lock. `withLock` will actually retry and semantically block

rtitle commented on google/src/main/scala/org/broadinstitute/dsde/workbench/util/DistributedLock.scala:
```diff
@@ -72,14 +73,14 @@ class DistributedLock[F[_]](lockPathPrefix: String, config: DistributedLockConfi
       data = Map(
```
This data is to implement a ttl on the lock, correct? (Just checking my understanding of this code)

Qi77Qi replied:
yep...

rtitle commented on google/src/test/scala/org/broadinstitute/dsde/workbench/google/Generators.scala:
```diff
@@ -0,0 +1,19 @@
+import java.time.Instant
+
+import org.broadinstitute.dsde.workbench.model.google._
+import org.scalacheck.Gen
```
Ooh scalacheck - lots of cool things in this PR :)

Qi77Qi replied:
I was thinking of adding scalacheck generators for wb-model too. I created some in sam repo https://github.com/broadinstitute/sam/blob/develop/src/test/scala/Generator.scala, but some of those probably can be moved to wb-model for sharing

rtitle commented on google/src/test/scala/org/broadinstitute/dsde/workbench/google/GoogleStorageInterpreterSpec.scala:
```diff
@@ -0,0 +1,107 @@
+// AsyncFlatSpec currently doesn't work with scalacheck's forAll. It'll be supported in scalatest 3
+class GoogleStorageInterpreterSpec extends AsyncFlatSpec with Matchers {
+  "ioStorage storeObject" should "be able to upload an object" in ioAssertion {
+    val bucketName = genGcsBucketName.sample.get
```
Interesting - is there a reason you're just taking 1 sample from the generators as opposed to something like:
```
forAll(genGcsBucketName, genGcsBlobName, genGcsObjectBody) { (bucketName, objectName, objectBody) =>
  // your assertion
}
```
which would generate many permutations of data and run the assertion each time? As written I feel like you’re not really using the power of ScalaCheck.

Qi77Qi replied:
see the comments on line 14 :)
I'm using `AsyncFlatSpec`. I wish it works with `forAll`

rtitle replied:
Ah ok, should have read the comment :)

rtitle commented on google/src/test/scala/org/broadinstitute/dsde/workbench/google/GoogleStorageInterpreterSpec.scala:
```diff
@@ -0,0 +1,107 @@
+object GoogleStorageInterpreterSpec {
+  implicit val cs = IO.contextShift(ExecutionContext.global)
+  implicit val timer = IO.timer(ExecutionContext.global)
+  val db = LocalStorageHelper.getOptions().getService()
```
Cool that Google provides this local in-memory implementation for testing.

Qi77Qi replied:
Yea, it's only for a few google services and they're not full featured

rtitle reviewed (changes requested):
This looks awesome! The tagless final approach and abstraction of the effect type works well here. The use of fs2 for retries and streaming of large data is nice too, I'd love to see that expand to other areas of workbench code.

Once it's finalized, it'd be nice to update Leo to use `GoogleStorageInterpreter` as well as Sam. We do things like create/delete buckets, set acls, and list objects. I think the Leo GCS code is mostly localized [here](https://github.com/DataBiosphere/leonardo/blob/develop/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/BucketHelper.scala).

I left some comments above, take them or leave them. My only real concern is with the implementation of `listObjectsWithPrefix` -- based on my reading of the javadoc it looks like it might return duplicate entries because it's streaming through pages and calling `iterateAll()` on each page. I'm not 100% sure of the behavior though -- may be worth a test with actual buckets.

Qi77Qi commented:
@rtitle  I'm not sure I understand the concern about duplicates completely. I added https://github.com/broadinstitute/workbench-libs/pull/188/files#diff-405a305f0d21ea56fc302bdd7467022bR101, I think duplicate entries are overwritten when storing objects.

rtitle commented:
^ elaborated in [this comment](https://github.com/broadinstitute/workbench-libs/pull/188#discussion_r237731810)

dvoet commented on google/src/main/scala/org/broadinstitute/dsde/workbench/google/GoogleStorageInterpreter.scala:
```diff
@@ -0,0 +1,119 @@
+  private val storageRetryableCode = Set(408, 429, 500, 502, 503, 504) //error codes are copied from com.google.cloud.storage.StorageException.RETRYABLE_ERRORS
+
+  val defaultRetryConfig = RetryConfig(
+    2 seconds,
```
I would start with 1 second and add a jitter of up to another second `org.broadinstitute.dsde.workbench.util.addJitter`

dvoet reviewed (commented):
1 more comment on retries then we need to figure out how to get the logs we want. That might not be actual code changes here but we should add it to the scaladoc or something.

danbills commented on google/src/test/scala/org/broadinstitute/dsde/workbench/google/GoogleStorageInterpreterSpec.scala:
```diff
@@ -0,0 +1,122 @@
+
+import scala.concurrent.ExecutionContext
+
+// AsyncFlatSpec currently doesn't work with scalacheck's forAll. It'll be supported in scalatest 3
```
scalatest 3 is out

Qi77Qi replied:
I realize the comment is a lie...It's really due to `forAll` is expecting a pure value...see https://github.com/rickynils/scalacheck/issues/214

danbills commented on google/src/main/scala/org/broadinstitute/dsde/workbench/google/GoogleStorageService.scala:
```diff
@@ -0,0 +1,41 @@
+  /**
+    * not memory safe
+    */
+  def unsafeGetObject(bucketName: GcsBucketName, blobName: GcsBlobName): F[Option[String]]
```
What is an `Object`?  i.e. can `String` b improved?  to me it looks like the contents of the file?  Also what's with the option, i.e. what does it mean when it's `None`

Qi77Qi replied:
`Object` here is just a blob. Return type is `String` because I used I piped it through `text.utf8Decode` before `.compile` it into `F`...Since lots of times for small file objects, `String` is probably what we want anyway. `None` means the blob doesn't exist in google.

danbills commented on google/src/main/scala/org/broadinstitute/dsde/workbench/google/GoogleStorageService.scala:
```diff
@@ -0,0 +1,41 @@
+    */
+  def unsafeListObjectsWithPrefix(bucketName: GcsBucketName, objectNamePrefix: String, maxPageSize: Long = 1000)(implicit sf: Sync[F]): F[List[GcsObjectName]] = listObjectsWithPrefix(bucketName, objectNamePrefix).compile.toList
+
+  def storeObject(bucketName: GcsBucketName, objectName: GcsBlobName, objectContents: Array[Byte], objectType: String): F[Unit]
```
consider using Scodec bitvector instead of `Array[Byte]`?  What about objectType, can `String` be improved?

Qi77Qi replied:
I don't know scodec well enough, but possibly...This PR is pretty big as is though, so I'll probably leave this as is and we can make future improvements.

danbills commented on google/src/main/scala/org/broadinstitute/dsde/workbench/google/GoogleStorageService.scala:
```diff
@@ -0,0 +1,41 @@
+  def unsafeGetObject(bucketName: GcsBucketName, blobName: GcsBlobName): F[Option[String]]
+
+  /**
+    * @return true if deleted; false if not found
```
IMO This should be an enumeration

danbills commented on google/src/main/scala/org/broadinstitute/dsde/workbench/google/GoogleStorageService.scala:
```diff
@@ -0,0 +1,41 @@
+  /**
+    * @return true if deleted; false if not found
+    */
+  def removeObject(bucketName: GcsBucketName, objectName: GcsBlobName): F[Boolean]
```
as above

danbills commented on google/src/main/scala/org/broadinstitute/dsde/workbench/google/GoogleStorageService.scala:
```diff
@@ -0,0 +1,41 @@
+  def createBucket(billingProject: GoogleProject, bucketName: GcsBucketName, acl: List[Acl]): F[Unit]
+}
+
+final case class GcsBlobName(value: String) extends AnyVal
```
this `String` could probably be refined via "Refined" library

Qi77Qi replied:
definitely worth improvements! In an other PR tho :)

danbills commented on google/src/main/scala/org/broadinstitute/dsde/workbench/google/util/DistributedLock.scala:
```diff
@@ -6,19 +6,23 @@ import java.util.concurrent.{Executor, TimeUnit}
-class DistributedLock[F[_]](lockPathPrefix: String, config: DistributedLockConfig, val googleFirestoreOps: GoogleFirestoreOps[F])(implicit ec: ExecutionContext, timer: Timer[F], af: Async[F]){
-
+class DistributedLock[F[_]](lockPathPrefix: String, config: DistributedLockConfig, val googleFirestoreOps: GoogleFirestoreService[F])(implicit ec: ExecutionContext, timer: Timer[F], asyncF: Async[F]) extends DistributedLockAlgebra[F]{
+  val retryable: Function1[Throwable, Boolean] = {
```
total nitpick but why not use syntax sugar for `Function1` type declaration `(Throwable) => Boolean`

Qi77Qi replied:
no reason other than I happened to be using `Function1`

danbills reviewed (commented):
Wondering why this isn't being used yet?  We could use it in Cromwell's DRS impl.

Lots of little questions but I really like the overall approach!

Qi77Qi commented:
I'm merging this PR for now but comments are still welcome and I'll address retrospectively

rtitle reviewed (commented):
Nice, 👍 . I like implementing it as a separate sub-module.
